In [14]:
import pandas as pd
import os
import xml
import numpy as np

# Data read and clean

In [102]:
corpusdocs = os.listdir("../corpus/")

In [135]:
alltex = pd.read_csv('alltexts copy.csv')

Select only those rows corresponding to Sumerian.

In [136]:
alltex = alltex.loc[alltex.lang == 'sux', :]

In [137]:
alltex['doc'] = pd.Series([str(i) for i in alltex.id_text])

In [140]:
alltex

Unnamed: 0         cf extent               form                 gw  \
0                0     dubsaŋ    NaN         dub-saŋ-ta              first   
1                1       Enki    NaN           {d}en-ki                  1   
2                2        unu    NaN               unu₂           dwelling   
3                3        gal    NaN                gal                big   
4                4         ed    NaN             im-ed₃             ascend   
5                5      anzag    NaN         an-zag-še₃            horizon   
6                6       anŋi    NaN             an-ŋi₆            eclipse   
7                7         zu    NaN                 zu               know   
8                8        ama    NaN                ama             mother   
9                9         tu    NaN                tu₆        incantation   
10              10         zu    NaN             zu-ke₄               know   
11              11         gi    NaN             ŋiš-gi            thicket   
12              12       tuku    NaN            tuku₄-e               rock   
13              13         AN    NaN                 AN                NaN   
14              14       KAŠ₄    NaN               KAŠ₄                NaN   
15              15         AN    NaN                 AN                NaN   
16              16       KAŠ₄    NaN               KAŠ₄                NaN   
17              17         me    NaN            me₃-ke₄             battle   
18              18     mašmaš    NaN            maš-maš           sorcerer   
19              19       erim    NaN              erim₂              enemy   
20              20        kur    NaN          kur₂-kur₂          different   
21              21   ŋiriŋena    NaN       ŋiri₃-ŋen-na               path   
22              22       Enki    NaN           {d}en-ki                  1   
23              23         ki    NaN                 ki              place   
24              24        unu    NaN               unu₂           dwelling   
25              25        gal    NaN                gal                big   
26              26         ed    NaN         im-ed₃-kam             ascend   
27              27        šag    NaN               šag₄              heart   
28              28        PU₂    NaN                PU₂                NaN   
29              29   1-kam-ma    NaN              1-kam                1st   
...            ...        ...    ...                ...                ...   
170826      170826         gu    NaN                gu₃              voice   
170827      170827         de    NaN                de₂               pour   
170828      170828      šinig    NaN     {ŋeš}šinig-am₆           tamarisk   
170829      170829       mete    NaN           me-te-bi  appropriate-thing   
170830      170830  TUR.UZ.ZA    NaN   TUR.UZ.ZA{mušen}              duck?   
170831      170831         gu    NaN                gu₃              voice   
170832      170832         de    NaN                de₂               pour   
170833      170833      ambar    NaN          ambar-am₃           reed-bed   
170834      170834       mete    NaN           me-te-bi  appropriate-thing   
170835      170835        šum    NaN   ga-ra-ab-šum₂-bi               give   
170836      170836         lu    NaN             lu₂-ra             person   
170837      170837         us    NaN       ga-ni-in-us₂             follow   
170838      170838      lugal    NaN              lugal               king   
170839      170839     dubsar    NaN            dub-sar             scribe   
170840      170840      lugal    NaN              lugal               king   
170841      170841        dur    NaN                dur               bond   
170842      170842        mah    NaN                mah              great   
170843      170843         id    NaN             id₂-da              river   
170844      170844      gilim    NaN             gib-ba              cross   
170845      170845         lu 

Now, the "terms" for the term freqeuency inverse doc frequency matrix need to be generated, so that's done here.

In [148]:
alltex['Term'] = [str(alltex.cf[i]) + "[" + str(alltex.gw[i]) + "]" + str(alltex.pos[i]) for i in alltex.index]
alltex

Unnamed: 0         cf extent               form                 gw  \
0                0     dubsaŋ    NaN         dub-saŋ-ta              first   
1                1       Enki    NaN           {d}en-ki                  1   
2                2        unu    NaN               unu₂           dwelling   
3                3        gal    NaN                gal                big   
4                4         ed    NaN             im-ed₃             ascend   
5                5      anzag    NaN         an-zag-še₃            horizon   
6                6       anŋi    NaN             an-ŋi₆            eclipse   
7                7         zu    NaN                 zu               know   
8                8        ama    NaN                ama             mother   
9                9         tu    NaN                tu₆        incantation   
10              10         zu    NaN             zu-ke₄               know   
11              11         gi    NaN             ŋiš-gi            thicket   
12              12       tuku    NaN            tuku₄-e               rock   
13              13         AN    NaN                 AN                NaN   
14              14       KAŠ₄    NaN               KAŠ₄                NaN   
15              15         AN    NaN                 AN                NaN   
16              16       KAŠ₄    NaN               KAŠ₄                NaN   
17              17         me    NaN            me₃-ke₄             battle   
18              18     mašmaš    NaN            maš-maš           sorcerer   
19              19       erim    NaN              erim₂              enemy   
20              20        kur    NaN          kur₂-kur₂          different   
21              21   ŋiriŋena    NaN       ŋiri₃-ŋen-na               path   
22              22       Enki    NaN           {d}en-ki                  1   
23              23         ki    NaN                 ki              place   
24              24        unu    NaN               unu₂           dwelling   
25              25        gal    NaN                gal                big   
26              26         ed    NaN         im-ed₃-kam             ascend   
27              27        šag    NaN               šag₄              heart   
28              28        PU₂    NaN                PU₂                NaN   
29              29   1-kam-ma    NaN              1-kam                1st   
...            ...        ...    ...                ...                ...   
170826      170826         gu    NaN                gu₃              voice   
170827      170827         de    NaN                de₂               pour   
170828      170828      šinig    NaN     {ŋeš}šinig-am₆           tamarisk   
170829      170829       mete    NaN           me-te-bi  appropriate-thing   
170830      170830  TUR.UZ.ZA    NaN   TUR.UZ.ZA{mušen}              duck?   
170831      170831         gu    NaN                gu₃              voice   
170832      170832         de    NaN                de₂               pour   
170833      170833      ambar    NaN          ambar-am₃           reed-bed   
170834      170834       mete    NaN           me-te-bi  appropriate-thing   
170835      170835        šum    NaN   ga-ra-ab-šum₂-bi               give   
170836      170836         lu    NaN             lu₂-ra             person   
170837      170837         us    NaN       ga-ni-in-us₂             follow   
170838      170838      lugal    NaN              lugal               king   
170839      170839     dubsar    NaN            dub-sar             scribe   
170840      170840      lugal    NaN              lugal               king   
170841      170841        dur    NaN                dur               bond   
170842      170842        mah    NaN                mah              great   
170843      170843         id    NaN             id₂-da              river   
170844      170844      gilim    NaN             gib-ba              cross   
170845      170845         lu 

In [150]:
pd.Series.unique(alltex.Term).shape

(4545,)

We see 4545 unique terms, and know that there are 394 docs, so expect to generate a 4545x394 pivot table.

Here, I will use a basic formula to calculate, as this is easily changed if necessary. 
First, note the formula
$$
tfidf(t, d, D) = tf(t,d)\cdot idf(t,D)
$$
where $t$ is the term, $d$ is the document, and $D$ is the corpus. 
Then, apply the raw count TF weight, and the inverse doc freq weight of 
$$
log\frac{N}{n_t}
$$
where $N = |D|$ and n_t is the number of docs where the term appears (formally $|{d \in D: t \in d}|$)


this means the overall expression for tfidf is
$$
f_{t,d}\cdot ln(\frac{N}{n_t})
$$

In [157]:
pivotcount = alltex.pivot_table(aggfunc=pd.Series.nunique, columns='doc', index='Term', values= 'form', dropna=False)
pivotcount

doc                   c.0.1.1  c.0.1.2  c.0.2.01  c.0.2.02  c.0.2.03  \
Term                                                                   
(X)[nan]nan               NaN      NaN       NaN       1.0       NaN   
1-kam-ma[1st]NU           1.0      NaN       NaN       NaN       NaN   
1/2[1/2]NU                NaN      NaN       NaN       NaN       NaN   
1/3[1/3]NU                NaN      NaN       NaN       NaN       NaN   
10-kam-ma[10th]NU         NaN      NaN       NaN       NaN       NaN   
1000[1000]NU              NaN      NaN       NaN       NaN       NaN   
100[100]NU                NaN      NaN       NaN       NaN       NaN   
108000[108000]NU          NaN      NaN       NaN       NaN       NaN   
1080[1080]NU              NaN      NaN       NaN       NaN       NaN   
108[108]NU                NaN      NaN       NaN       NaN       NaN   
10[10]NU                  NaN      1.0       NaN       NaN       NaN   
11-kam-ma[11th]NU         NaN      NaN       NaN       NaN       NaN   
110[110]NU                NaN      NaN       NaN       NaN       NaN   
1110[1110]NU              NaN      NaN       NaN       NaN       NaN   
116[116]NU                NaN      NaN       NaN       NaN       NaN   
117[117]NU                NaN      NaN       NaN       NaN       NaN   
11[11]NU                  NaN      NaN       NaN       NaN       NaN   
12-kam-ma[12th]NU         NaN      NaN       NaN       NaN       NaN   
1200[1200]NU              NaN      NaN       NaN       NaN       NaN   
120[120]NU                NaN      NaN       NaN       NaN       NaN   
123[123]NU                NaN      NaN       NaN       NaN       NaN   
124[124]NU                NaN      NaN       NaN       NaN       NaN   
125[125]NU                NaN      NaN       NaN       NaN       NaN   
126[126]NU                NaN      NaN       NaN       NaN       NaN   
12[12]NU                  NaN      NaN       NaN       NaN       NaN   
131[131]NU                NaN      NaN       NaN       NaN       NaN   
1320[1320]NU              NaN      NaN       NaN       NaN       NaN   
132[132]NU                NaN      NaN       NaN       NaN       NaN   
134[134]NU                NaN      NaN       NaN       NaN       NaN   
136[136]NU                NaN      NaN       NaN       NaN       NaN   
...                       ...      ...       ...       ...       ...   
šulalum[punishment]N      NaN      NaN       NaN       NaN       NaN   
šulu[bird]N               NaN      NaN       NaN       NaN       NaN   
šuluh[cleansing]N         NaN      NaN       NaN       NaN       NaN   
šum[garlic]N              NaN      NaN       NaN       NaN       NaN   
šum[give]V/t              NaN      NaN       NaN       NaN       NaN   
šum[slaughter]V/t         NaN      NaN       NaN       NaN       NaN   
šumhuš[vegetable]N        NaN      NaN       NaN       NaN       NaN   
šumin[bar]N               NaN      NaN       NaN       NaN       NaN   
šumin[stone]N             NaN      NaN       NaN       NaN       NaN   
šumme[saw]N               NaN      NaN       NaN       NaN       NaN   
šumsikil[vegetable]N      NaN      NaN       NaN       NaN       NaN   
šumunša[song]N            NaN      NaN       NaN       NaN       NaN   
šunim[spring]N            NaN      NaN       NaN       NaN       NaN   
šunir[emblem]N            NaN      NaN       NaN       NaN       NaN   
šuniŋin[total]N           NaN      1.0       NaN       NaN       NaN   
šur[eyebrow]N             NaN      NaN       NaN       NaN       NaN   
šurin[oven]N              NaN      NaN       NaN       NaN       NaN   
šurmen[cypress]N          NaN      NaN       NaN       NaN       NaN   
šurum[litter]N            NaN      NaN       NaN       NaN       NaN   
šusar[string]N            NaN      NaN       NaN       NaN       NaN   
šusi[finger]N             NaN      NaN       NaN       NaN       NaN   
šuteŋ[accept]V/t          NaN      NaN       NaN       NaN       NaN   
šutubur[mixture]N         NaN      NaN  

Terms that are in a list of unique values for Term column, but did not show up in pivot table.

In [165]:
notin = pd.Series.unique(alltex.Term)[~np.in1d(pd.Series.unique(alltex.Term), pivotcount.index)]

In [169]:
alltex.loc[np.in1d(alltex.Term, notin), :]

Unnamed: 0                 cf extent                      form  \
167540      167540            namaŋal    NaN          nam-a₂-ŋal₂-ŋu₁₀   
167543      167543            damegir    NaN            dam-eŋer-ra-ni   
167590      167590                sar    NaN              na-an-sar-re   
167657      167657            aʾambar    NaN             a-ambar-ra-am   
167700      167700                par    NaN             i₃-par₃-re-eš   
167707      167707                 MI    NaN                  nu-ki-MI   
167720      167720                 MI    NaN                     ga-MI   
167728      167728              GA-GA    NaN                     GA-GA   
167732      167732             saŋ.DU    NaN                    saŋ-DU   
167905      167905                ula    NaN                     u₃-la   
167907      167907                ula    NaN                     u₃-la   
168138      168138              azbal    NaN               {ŋeš}az-bal   
168288      168288                 UR    NaN               nam-ba-UR-e   
168735      168735           namušbar    NaN      nam-ušbar-ra-ŋa₂-ke₄   
168750      168750            aʾambar    NaN                a-ambar-ra   
168789      168789               HUB₂    NaN           mu-un-HUB₂-HUB₂   
169273      169273            daʾerua    NaN                 da-e-ru-a   
169276      169276               sugu    NaN                    su-gu₇   
169281      169281            daʾerua    NaN                   da-ri-a   
169282      169282              SUHUR    NaN                     SUHUR   
169286      169286               sugu    NaN                    su-gu₇   
169313      169313               a-KU    NaN                      a-KU   
169428      169428               amma    NaN                    am₃-ma   
169565      169565               kiʾe    NaN                  ki-e₃-bi   
169747      169747              ebara    NaN                 e₂-bar-ra   
169748      169748              eškiŋ    NaN           eš-kiŋ₂-zu-gin₇   
169762      169762                GIG    NaN                       GIG   
170107      170107             šubila    NaN       {gi}šu-bil-la₂-ŋu₁₀   
170147      170147          hizurbara    NaN      hi-iz-ur-bar-ra{sar}   
170148      170148         ŋeštinkaʾa    NaN     ŋeštin-ka₅-a{sar}-ke₄   
170149      170149                 ug    NaN                    {u₂}ug   
170150      170150           piriŋ.GI    NaN              {u₂}piriŋ-GI   
170151      170151        emeʾurgirak    NaN   {u₂}eme-ur-gir₁₅-ra-ke₄   
170152      170152            niŋurum    NaN           {u₂}niŋ₂-u₂-rum   
170153      170153             guʾira    NaN         {u₂}gu₃-i₃-ra-ke₄   
170475      170475              šudib    NaN                    šu-dib   
170535      170535               MAŠ₂    NaN                      MAŠ₂   
170671      170671           niŋtukum    NaN        niŋ₂-tukum-ta-e-še   
170811      170811               adlu    NaN             ad₄-lu₂-u₃-ka   
170816      170816            kizurra    NaN             ki-zu₂-ur₅-ra   
170825      170825  KU.GA₂.NUN.tir.ra    NaN  KU.ŊA₂.NUN-tir-ra{mušen}   
170830      170830          TUR.UZ.ZA    NaN          TUR.UZ.ZA{mušen}   

                     gw  id_text lang line_no  line_ref  pos status  \
167540         strength  c.6.2.1  sux       3       291    N    NaN   
167543             wife  c.6.2.1  sux       1       292    N    NaN   
167590            smoke  c.6.2.1  sux       1       307  V/t    NaN   
167657            water  c.6.2.2  sux       2         2    N    NaN   
167700            unmng  c.6.2.2  sux       4        12    V    NaN   
167707               MI  c.6.2.2  sux       1        14    V    NaN   
167720               MI  c.6.2.2  sux       4        17    V    NaN   
167728              NaN  c.6.2.2  sux       2        19  NaN    NaN   
167732   smitten-person  c.6.2.2  sux       3        20    N    NaN   
167905         anything  c.6.2.3  sux       4         8    N    NaN   
167907         anything  c.6.2.3  sux       

We now see that these terms weren't included in the pivot table because they have no doc entry (thus id_text was probably also NaN).  Let's try it with titles instead.

In [171]:
pivottitlescount = alltex.pivot_table(aggfunc=pd.Series.nunique, columns='text_name', index='Term', values= 'form', dropna=False)
pivottitlescount

text_name              A praise poem of Šulgi (Šulgi P)  \
Term                                                      
(X)[nan]nan                                         1.0   
1-kam-ma[1st]NU                                     NaN   
1/2[1/2]NU                                          NaN   
1/3[1/3]NU                                          NaN   
10-kam-ma[10th]NU                                   NaN   
1000[1000]NU                                        NaN   
100[100]NU                                          NaN   
108000[108000]NU                                    NaN   
1080[1080]NU                                        NaN   
108[108]NU                                          NaN   
10[10]NU                                            NaN   
11-kam-ma[11th]NU                                   NaN   
110[110]NU                                          NaN   
1110[1110]NU                                        NaN   
116[116]NU                                          NaN   
117[117]NU                                          NaN   
11[11]NU                                            NaN   
12-kam-ma[12th]NU                                   NaN   
1200[1200]NU                                        NaN   
120[120]NU                                          NaN   
123[123]NU                                          NaN   
124[124]NU                                          NaN   
125[125]NU                                          NaN   
126[126]NU                                          NaN   
12[12]NU                                            NaN   
131[131]NU                                          NaN   
1320[1320]NU                                        NaN   
132[132]NU                                          NaN   
134[134]NU                                          NaN   
136[136]NU                                          NaN   
...                                                 ...   
šulalum[punishment]N                                NaN   
šulu[bird]N                                         NaN   
šuluh[cleansing]N                                   NaN   
šum[garlic]N                                        NaN   
šum[give]V/t                                        1.0   
šum[slaughter]V/t                                   NaN   
šumhuš[vegetable]N                                  NaN   
šumin[bar]N                                         NaN   
šumin[stone]N                                       NaN   
šumme[saw]N                                         NaN   
šumsikil[vegetable]N                                NaN   
šumunša[song]N                                      NaN   
šunim[spring]N                                      NaN   
šunir[emblem]N                                      NaN   
šuniŋin[total]N                                     NaN   
šur[eyebrow]N                                       NaN   
šurin[oven]N                                        NaN   
šurmen[cypress]N                                    NaN   
šurum[litter]N                                      NaN   
šusar[string]N                                      NaN   
šusi[finger]N                                       NaN   
šuteŋ[accept]V/t                                    NaN   
šutubur[mixture]N                                   NaN   
šutug[reed-hut]N                                    NaN   
šutum[storehouse]N                                  NaN   
šutur[garment]N                                     NaN   
šuš[cover]V/t                                       NaN   
šuʾi[barber]N                                       NaN   
šuʾu[stone]N                                        NaN   
šuʾura[goose]N                                      NaN   

text_name             A balbale (?) to  Inana (Inana F)  \
Term                                                      
(X)[nan]nan                                         NaN   
1-kam-ma[1st]NU                                     NaN   
1/2[1/2]NU                                    

This looks more complete, with all 4545 terms accounted for. Now, we have count per document, so each of these cells is a particular $f_{t,d}$

To calculate log(N/n_t), we can simply note that N is 394, and n_t is the number of not NaN entries per row. This is easily done.

In [183]:
def idf(row):  
    return row * np.sum(~np.isnan(pivottitlescount.iloc[1]))/len(row)

In [185]:
tfidftable = pivottitlescount.apply(func=idf)

Replace NaN with 0, technically that makes more sense for term frequency number.

In [187]:
tfidftable = tfidftable.fillna(value = 0)

text_name              A praise poem of Šulgi (Šulgi P)  \
Term                                                      
(X)[nan]nan                                     0.00374   
1-kam-ma[1st]NU                                 0.00000   
1/2[1/2]NU                                      0.00000   
1/3[1/3]NU                                      0.00000   
10-kam-ma[10th]NU                               0.00000   
1000[1000]NU                                    0.00000   
100[100]NU                                      0.00000   
108000[108000]NU                                0.00000   
1080[1080]NU                                    0.00000   
108[108]NU                                      0.00000   
10[10]NU                                        0.00000   
11-kam-ma[11th]NU                               0.00000   
110[110]NU                                      0.00000   
1110[1110]NU                                    0.00000   
116[116]NU                                      0.00000   
117[117]NU                                      0.00000   
11[11]NU                                        0.00000   
12-kam-ma[12th]NU                               0.00000   
1200[1200]NU                                    0.00000   
120[120]NU                                      0.00000   
123[123]NU                                      0.00000   
124[124]NU                                      0.00000   
125[125]NU                                      0.00000   
126[126]NU                                      0.00000   
12[12]NU                                        0.00000   
131[131]NU                                      0.00000   
1320[1320]NU                                    0.00000   
132[132]NU                                      0.00000   
134[134]NU                                      0.00000   
136[136]NU                                      0.00000   
...                                                 ...   
šulalum[punishment]N                            0.00000   
šulu[bird]N                                     0.00000   
šuluh[cleansing]N                               0.00000   
šum[garlic]N                                    0.00000   
šum[give]V/t                                    0.00374   
šum[slaughter]V/t                               0.00000   
šumhuš[vegetable]N                              0.00000   
šumin[bar]N                                     0.00000   
šumin[stone]N                                   0.00000   
šumme[saw]N                                     0.00000   
šumsikil[vegetable]N                            0.00000   
šumunša[song]N                                  0.00000   
šunim[spring]N                                  0.00000   
šunir[emblem]N                                  0.00000   
šuniŋin[total]N                                 0.00000   
šur[eyebrow]N                                   0.00000   
šurin[oven]N                                    0.00000   
šurmen[cypress]N                                0.00000   
šurum[litter]N                                  0.00000   
šusar[string]N                                  0.00000   
šusi[finger]N                                   0.00000   
šuteŋ[accept]V/t                                0.00000   
šutubur[mixture]N                               0.00000   
šutug[reed-hut]N                                0.00000   
šutum[storehouse]N                              0.00000   
šutur[garment]N                                 0.00000   
šuš[cover]V/t                                   0.00000   
šuʾi[barber]N                                   0.00000   
šuʾu[stone]N                                    0.00000   
šuʾura[goose]N                                  0.00000   

text_name             A balbale (?) to  Inana (Inana F)  \
Term                                                      
(X)[nan]nan                                         0.0   
1-kam-ma[1st]NU                                     0.0   
1/2[1/2]NU                                    